### ISSUE TITLE
threading returning too early

### ISSUE THREAD

**Issue by PaulSchweizer** [PaulSchweizer/flowpipe#93]
We are now returning too early from the threading function because we remove the nodes from the nodes we check right after starting the thread: https://github.com/PaulSchweizer/flowpipe/blob/master/flowpipe/graph.py#L182

That way this is automatically true after having started the last thread: https://github.com/PaulSchweizer/flowpipe/blob/master/flowpipe/graph.py#L187

This could be a solution:

```python
def _evaluate_threaded(self, skip_clean, submission_delay, raise_after,
                       **kwargs):
    threads = {}
    nodes_to_evaluate = [n for n in self.evaluation_sequence
                         if n.is_dirty or not skip_clean]
    nodes_to_check = [n for n in self.evaluation_sequence
                      if n.is_dirty or not skip_clean]
    empty_loops = 0
    while True:
        for node in nodes_to_evaluate:
            if (node.name not in threads
                    and all(not n.is_dirty for n in node.upstream_nodes)):
                # If all deps are ready and no thread is active, create one
                threads[node.name] = threading.Thread(
                    target=node.evaluate,
                    name="flowpipe.{0}.{1}".format(self.name, node.name))
                threads[node.name].start()
                nodes_to_evaluate.remove(node)

        graph_threads = [t for t in threading.enumerate()
                         if t.name.startswith(
                             "flowpipe.{0}".format(self.name))]
        all_clean = all(not n.is_dirty for n in nodes_to_check)
        if len(graph_threads) == 0 and not all_clean:  # pragma: no cover
            # No more threads running after a round of submissions means
            # we're either done or stuck
            if raise_after is not None and empty_loops > raise_after:
                raise RuntimeError(
                    "Could not sucessfully compute all nodes in the "
                    "graph {0}".format(self.name))
            else:
                empty_loops += 1
        else:
            empty_loops = 0

        if all_clean:
            break
        time.sleep(submission_delay)
```

**Comment by neuneck** [comment]
Thank you for finding this. I like the solution, too.

**Comment by PaulSchweizer** [comment]
Resolved and released to PyPI as v0.5.6
